serverless/question-service/functions/api/questions-content.ts

```diff
@@ -0,0 +1,157 @@
+import { PrismaClient } from "@prisma/client";
+import axios from "axios";
+import { logger } from "firebase-functions";
+import { isEmpty } from "lodash";
+
+import {
+  getQuestionFromSlug,
+  LEETCODE_GRAPHQL_ENDPOINT,
+} from "../constants/graphql.queries";
+import {
+  LeetcodeContentType,
+  NormalisedQuestionContentType,
+  QuestionContentResponse,
+} from "../constants/graphql.types";
+import { processConcurrently } from "../utils/concurrency";
+
+// ***** Prisma ***** //
+
+export async function getExistingQuestionContents(prisma: PrismaClient) {
+  const contentMap: Record<string, NormalisedQuestionContentType> = {};
+  const dbQuestionContent = await prisma.questionContent.findMany({
+    include: { hints: true },
+  });
+
+  if (isEmpty(dbQuestionContent)) {
+    logger.info("content database is empty");
+    return dbQuestionContent;
+  }
+
+  for (const { titleSlug, content, hints } of dbQuestionContent) {
+    const hintMap: Record<number, string> = {};
+
+    for (const { hintId, hint } of hints) {
+      if (hint) {
+        hintMap[hintId] = hint;
+      }
+    }
+
+    contentMap[titleSlug] = {
+      content,
+      hints,
+      titleSlug,
+    };
+  }
+
+  return contentMap;
+}
+
+export async function insertQuestionContent(
+  prisma: PrismaClient,
+  newSlugs: string[]
+) {
+  try {
+    const lcData = await getLcQuestionContent(newSlugs);
+    for (const { content, hints, titleSlug } of Object.values(lcData)) {
+      await prisma.questionContent.create({
+        data: {
+          content,
+          hints: {
+            create: hints.map((v) => {
+              return { hint: v };
+            }),
+          },
+          titleSlug: titleSlug.toLowerCase(),
+        },
+      });
+    }
+  } catch (error) {
+    logger.error(error);
+  }
+}
+
+export async function purgeStaleQuestionContent(
+  prisma: PrismaClient,
+  staleContentSet: string[]
+) {
+  if (staleContentSet.length == 0) {
+    return;
+  }
+
+  try {
+    await prisma.questionContent.deleteMany({
+      where: {
+        titleSlug: {
+          in: staleContentSet,
+        },
+      },
+    });
+  } catch (error) {
+    logger.error(error);
+  }
+}
+
+// ***** GraphQL-LeetCode ***** //
+
+const axiosInstance = axios.create({
+  headers: {
+    "Content-Type": "application/json",
+  },
+});
+
+async function getLcQuestionContent(titleSlugs: string[]) {
+  const lcMap: Record<string, LeetcodeContentType> = {};
+  const getLcContentFuncs: (() => Promise<void>)[] = [];
+  for (const titleSlug of titleSlugs) {
+    getLcContentFuncs.push(async () => {
+      try {
+        const { data } = await axiosInstance.post<QuestionContentResponse>(
+          LEETCODE_GRAPHQL_ENDPOINT,
+          getQuestionFromSlug(titleSlug)
+        );
+        const { content, hints } = data.data.question;
+        lcMap[titleSlug] = {
+          content,
+          hints,
+          titleSlug,
+        };
+      } catch (error) {}
+    });
+  }
```

Comment:
can be written to be more readable using a `map`:
```suggestion
const getLcdContents = titleSlugs.map((titleSlug) => {
    return async () => {
      try {
        const { data } = await axiosInstance.post<QuestionContentResponse>(
          LEETCODE_GRAPHQL_ENDPOINT,
          getQuestionFromSlug(titleSlug)
        );
        const { content, hints } = data.data.question;
        lcMap[titleSlug] = {
          content,
          hints,
          titleSlug,
        };
      } catch (error) {}
    }
  });
```

prioritize using `map` instead of `for` in general - usually results in cleaner code. also dont think we catching any errors here? what happens if an error occurs ah does the code j hang/break?

Reply:
error-handling is a grey area in FP which is why I avoided using `map` since its mainly used for deterministic transformations since it can cause side-effects, so I opted for a for-loop and in this instance i feel like the for-loop is more readable. 

it's not really accurate to call it `getLcdContents` as its an array of async functions that would later on be resolved in `processConcurrently` to get the LC content

errors that are thrown are ignored as it's more likely that the error is caused by getting rate-limited by LC. this only affects us when we're seeding the db since LC adds about a handful of questions at a given time which we can resolve without getting rate-limited.

Reply:
so main reason for not using `map` in this context is that error-handling is a grey area in FP? i think our situation here is v similar to [this](https://stackoverflow.com/a/45028737/18452483) which uses a solution similar to the one i suggested